numpy/core/shape_base.py

```diff
@@ -417,38 +419,100 @@ def format_index(index):
         return first_index, max_arr_ndim
     elif type(arrays) is list and len(arrays) == 0:
         # We've 'bottomed out' on an empty list
-        return parent_index + [None], 0
+        raise ValueError('Lists cannot be empty. '
+                         'An empty list was found at index {}'
+                         ''.format(format_index(parent_index)))
     else:
         # We've 'bottomed out' - arrays is either a scalar or an array
         return parent_index, _nx.ndim(arrays)
 
 
-def _block(arrays, max_depth, result_ndim):
-    """
-    Internal implementation of block. `arrays` is the argument passed to
-    block. `max_depth` is the depth of nested lists within `arrays` and
-    `result_ndim` is the greatest of the dimensions of the arrays in
-    `arrays` and the depth of the lists in `arrays` (see block docstring
-    for details).
+def _concatenate_shapes(shapes, axis):
+    """Given array shapes, return the resulting shape that would occur
+    after array concatenation.
+
+    concatenate(arrs, axis).shape == _concatenate_shapes([a.shape for a in arrs], axis)
     """
-    def atleast_nd(a, ndim):
-        # Ensures `a` has at least `ndim` dimensions by prepending
-        # ones to `a.shape` as necessary
-        return array(a, ndmin=ndim, copy=False, subok=True)
-
-    @recursive
-    def block_recursion(self, arrays, depth=0):
-        if depth < max_depth:
-            if len(arrays) == 0:
-                raise ValueError('Lists cannot be empty')
-            arrs = [self(arr, depth+1) for arr in arrays]
-            return _nx.concatenate(arrs, axis=-(max_depth-depth))
-        else:
-            # We've 'bottomed out' - arrays is either a scalar or an array
-            # type(arrays) is not list
-            return atleast_nd(arrays, result_ndim)
+    # Take a shape, any shape
+    first_shape = shapes[0]
+    for shape in shapes[1:]:
+        if (shape[:axis] != first_shape[:axis] or
+                shape[axis+1:] != first_shape[axis+1:]):
+            raise ValueError('Mismatched array shapes in block.')
+    shape_on_dim = sum(shape[axis] for shape in shapes)
+    return first_shape[:axis] + (shape_on_dim,) + first_shape[axis+1:]
+
+
+def _atleast_nd(a, ndim):
+    # Ensures `a` has at least `ndim` dimensions by prepending
+    # ones to `a.shape` as necessary
+    return array(a, ndmin=ndim, copy=False, subok=True)
+
+
+def _accumulate(values):
+    # Helper function because Python 2.7 doesn't have
+    # itertools.accumulate
+    value = 0
+    accumulated = []
+    for v in values:
+        value += v
+        accumulated.append(value)
 
-    return block_recursion(arrays)
+    return tuple(accumulated)
+
+
+def _block_info_recursion(arrays, list_ndim, result_ndim, depth=0):
+    """Form required information to block the matricies using a flat iterator.
```

Comment:
Lets avoid the word `matrices` or any spellings thereof, since that's associated with 2d and `np.matrix`. "arrays" would be preferable